src/librustc_resolve/diagnostics.rs

```diff
@@ -857,54 +858,173 @@ impl<'a> Resolver<'a> {
                     ));
                 }
+            } else {
+                let extern_prelude_names = self.extern_prelude.clone();
+                for (ident, _) in extern_prelude_names.into_iter() {
+                    if ident.span.from_expansion() {
+                        // Idents are adjusted to the root context before being
+                        // resolved in the extern prelude, so reporting this to the
+                        // user is no help. This skips the injected
+                        // `extern crate std` in the 2018 edition, which would
+                        // otherwise cause duplicate suggestions.
+                        continue;
+                    }
+                    if let Some(crate_id) = self.crate_loader.maybe_process_path_extern(ident.name)
+                    {
+                        crate_ids.push((crate_id, ident));
+                    }
+                }
+            }
+
+            for (crate_id, ident) in crate_ids.into_iter() {
+                let crate_root = self.get_module(DefId { krate: crate_id, index: CRATE_DEF_INDEX });
+
+                suggestions.extend(self.lookup_import_candidates_from_module(
+                    lookup_ident,
+                    namespace,
+                    parent_scope,
+                    crate_root,
+                    ident,
+                    &filter_fn,
+                    &mut collect_fn,
+                ));
             }
         }
 
         suggestions
     }
 
+    /// The purpose of this function is to collect public symbols names that are unique across all
+    /// crates in the build. Later, when printing about types we can use those names instead of the
+    /// full exported path to them.
+    ///
+    /// So essentially, if a symbol name can only be imported from one place for a type, and as
+    /// long as it was not glob-imported anywhere in the current crate, we can trim its printed
+    /// path and print only the name.
+    ///
+    /// This has wide implications on error messages with types, for example, shortening
+    /// `std::vec::Vec` to just `Vec`, as long as there is no other `Vec` importable anywhere.
+    ///
+    /// The implementation uses that same underyling mechanism that 'use' suggestions are using,
+    /// and there a few aspects to how this is done:
+    ///
+    /// * Ignore type aliases. This allows `std::result::Result` to compete over the
+    ///   `std::io::Result` alias.
+    /// * Separate uniqueness for trait symbols and type symbols. This sorts out an overlap between
+    ///   `std::fmt::Display` and `std::path::Display`, for instance.
+    /// * Skip Variant types. Othersize, it can pick up `core::num::flt2dev::Part::Copy` as collsion
+    ///   with `std::marker::Copy`. It also makes sense, as users cannot specifically import this
+    ///   into the Type namespace (rather than the Value namespace).
```

Comment:
Looks like you want to special case some well known names (i.e. `Copy` is always trimmed), but do that with more general heuristics that could be wrong for some other user-defined items that are not well-known.

I would suggest to get rid of all these heuristics in the initial PR (and remove `enum Kind`).
If two items with the same name have different `DefIds`, the I suggest to consider them non-unique.
The heuristics can be submitted as separate PRs later, then their impact could be reviewed better.

Reply:
Done - separated away from this PR, and prepared another commit for later.

Reply:
(waiting in a side-branch: https://github.com/rust-lang/rust/commit/fb92566fd53f89dc158096e242bc5578a3971d42)